### PR TITLE
Fix and unify artifact collection in CI script

### LINKED ISSUE #1910: E2E artifacts are not collected in CI
### What happened?

https://github.com/scylladb/scylla-operator/pull/1754 accidentally broke e2e artifact collection in our CI script. The e2e artifacts are not being collected.

See e.g. https://gcsweb.scylla-operator.scylladb.com/gcs/scylla-operator-prow/pr-logs/pull/scylladb_scylla-operator/1850/pull-scylla-operator-master-e2e-gke-parallel/1787741759061626880/artifacts/.

/priority critical-urgent
/kind bug
/assign

### What did you expect to happen?

e2e artifacts should be collected along with must-gather

### How can we reproduce it (as minimally and precisely as possible)?

Look into any CI run post-merge.

### Scylla Operator version

master

### Kubernetes platform name and version

n/a

### Please attach the must-gather archive.

n/a

### Anything else we need to know?

n/a

### PR DESCRIPTION
**Pull request by rzetelskik**
**Description of your changes:** https://github.com/scylladb/scylla-operator/pull/1754 broke e2e artifacts collection in our CI. This PR fixes it by waiting for the e2e container to exit and then collecting the artifacts from its neighbour. It also removes the superfluous concurrency and unifies the collection of must-gather artifacts with e2e artifacts.

**Which issue is resolved by this Pull Request:**
Resolves #1910 

/kind bug
/priority critical-urgent

### CODE REVIEW

File: hack/.ci/run-e2e-gke.sh [hunk 1]
```diff
@@ -54,32 +54,64 @@ function kubectl_create {
 }
 
 function gather-artifacts {
-  kubectl -n=e2e run --restart=Never --image="${SO_IMAGE}" --labels='app=must-gather' --command=true must-gather -- bash -euExo pipefail -O inherit_errexit -c "function wait-for-artifacts { touch /tmp/done && until [[ -f '/tmp/exit' ]]; do sleep 1; done } && trap wait-for-artifacts EXIT && mkdir /tmp/artifacts && scylla-operator must-gather --all-resources --loglevel=2 --dest-dir=/tmp/artifacts"
+  kubectl_create -n=e2e -f=- <<EOF
+  apiVersion: v1
```
Comment (tnozicka) [inline hack/.ci/run-e2e-gke.sh:58]: nit: EOF wise, I think this should start at column 0

File: hack/.ci/run-e2e-gke.sh [hunk 2]
```diff
@@ -54,32 +54,64 @@ function kubectl_create {
...
+    containers:
+    - name: wait-for-artifacts
+      command:
+      - bash
+      - -euExo
+      - pipefail
+      - -O
+      - inherit_errexit
+      - -c
+      - sleep infinity
```
Comment (tnozicka) [inline hack/.ci/run-e2e-gke.sh:75]:
```suggestion
      - /usr/bin/sleep
      - infinity
```

File: hack/.ci/run-e2e-gke.sh [hunk 3]
```diff
@@ -54,32 +54,64 @@ function kubectl_create {
...
+      image: "${SO_IMAGE}"
+      imagePullPolicy: Always
+      volumeMounts:
+      - name: artifacts
+        mountPath: /tmp/artifacts
+    - name: must-gather
+      command:
+      - scylla-operator
+      args:
```
Comment (tnozicka) [inline hack/.ci/run-e2e-gke.sh:84]: optional nit: if you are changing command, I'd say you can specify the options there. separate args are mean for use with a "default" command)

File: hack/.ci/run-e2e-gke.sh [hunk 4]
```diff
@@ -54,32 +54,64 @@ function kubectl_create {
...
-    ls -l "${ARTIFACTS}"
-  ) &
-  must_gather_bg_pid=$!
-
-  kubectl -n=e2e logs -f pod/must-gather
-  exit_code=$( kubectl -n=e2e get pods/must-gather --output='jsonpath={.status.containerStatuses[0].state.terminated.exitCode}' )
+  exit_code=""
+  while [[ "${exit_code}" == "" ]]; do
+    kubectl -n=e2e logs -f pod/must-gather -c=must-gather || echo "kubectl logs failed before pod has finished, retrying..." > /dev/stderr
+    exit_code="$( kubectl -n=e2e get pods/must-gather --template='{{ range .status.containerStatuses }}{{ if and (eq .name "must-gather") (ne .state.terminated.exitCode nil) }}{{ .state.terminated.exitCode }}{{ end }}{{ end }}' )"
```
Comment (tnozicka) [inline hack/.ci/run-e2e-gke.sh:103]: lets make this a function and reuse it

File: hack/.ci/run-e2e-gke.sh [hunk 5]
```diff
@@ -200,16 +232,7 @@ spec:
...
-    - |
-      function wait-for-artifacts {
-        touch /tmp/done
-        until [[ -f '/tmp/exit' ]]; do
-          sleep 1
-        done
-      }
-
-      trap wait-for-artifacts EXIT
+    - sleep infinity
```
Comment (tnozicka) [inline hack/.ci/run-e2e-gke.sh:235]: and here

File: hack/.ci/run-e2e-gke.sh (hunk 1 above)
Comment (rzetelskik) [reply hack/.ci/run-e2e-gke.sh]: done

File: hack/.ci/run-e2e-gke.sh (hunk 2 above)
Comment (rzetelskik) [reply hack/.ci/run-e2e-gke.sh]: done

File: hack/.ci/run-e2e-gke.sh (hunk 3 above)
Comment (rzetelskik) [reply hack/.ci/run-e2e-gke.sh]: removed the command since it's the default

File: hack/.ci/run-e2e-gke.sh (hunk 5 above)
Comment (rzetelskik) [reply hack/.ci/run-e2e-gke.sh]: done

File: hack/.ci/run-e2e-gke.sh [hunk 6]
```diff
@@ -53,33 +53,66 @@ function kubectl_create {
     fi
 }
 
+# $1 - namespace
+# $2 - pod name
+# $3 - container name
+function await-exit-code {
```
Comment (tnozicka) [inline hack/.ci/run-e2e-gke.sh:59]:
```suggestion
function wait-for-container-exit-with-logs {
```

**Review by tnozicka — APPROVED** [review]
/approve
/lgtm
thanks for fixing this

/hold
to check artifacts manually when the jobs finish

File: hack/.ci/run-e2e-gke.sh [hunk 7]
```diff
@@ -53,33 +53,66 @@ function kubectl_create {
...
+# $1 - namespace
+# $2 - pod name
+# $3 - container name
+function wait-for-container-exit-with-logs {
+  exit_code=""
+  while [[ "${exit_code}" == "" ]]; do
+    kubectl -n="${1}" logs -f pod/"${2}" -c="${3}" || echo "kubectl logs failed before pod has finished, retrying..." > /dev/stderr
+    exit_code="$( kubectl -n="${1}" get pods/"${2}" --template='{{ range .status.containerStatuses }}{{ if and (eq .name "'"${3}"'") (ne .state.terminated.exitCode nil) }}{{ .state.terminated.exitCode }}{{ end }}{{ end }}' )"
+  done
+  echo "${exit_code}"
```
[the remark this answers is not in the document]
Comment (rzetelskik) [reply hack/.ci/run-e2e-gke.sh]: fixed, thx

**Review by tnozicka — COMMENTED** [review]
/lgtm

**Comment by tnozicka** [comment]
/lgtm cancel
I've tried to pick this into my PR but the exit code is incorrectly set even here
```
 + exit_code='I0507 12:39:52.777049       1 featuregate/feature_gate.go:249] feature gates: &{map[AllAlpha:true AllBeta:true AutomaticTLSCertificates:true]}
I0507 12:39:52.777228       1 tests/tests.go:30] maxprocs: Leaving GOMAXPROCS=[8]: CPU quota undefined 
```
those logs should be on stderr :(
https://prow.scylla-operator.scylladb.com/view/gs/scylla-operator-prow/pr-logs/pull/scylladb_scylla-operator/1911/pull-scylla-operator-master-e2e-gke-serial/1787821240048685056#1:test-build-log.txt%3A524

File: hack/.ci/run-e2e-gke.sh [hunk 8]
```diff
@@ -53,33 +53,66 @@ function kubectl_create {
     fi
 }
 
+# $1 - namespace
+# $2 - pod name
+# $3 - container name
+function wait-for-container-exit-with-logs {
+  exit_code=""
+  while [[ "${exit_code}" == "" ]]; do
+    kubectl -n="${1}" logs -f pod/"${2}" -c="${3}" || echo "kubectl logs failed before pod has finished, retrying..." > /dev/stderr
```
Comment (tnozicka) [inline hack/.ci/run-e2e-gke.sh:62]:
```suggestion
    kubectl -n="${1}" logs -f pod/"${2}" -c="${3}" > /dev/stderr || echo "kubectl logs failed before pod has finished, retrying..." > /dev/stderr
```

**Comment by tnozicka** [comment]
/lgtm
thanks

**Comment by rzetelskik** [comment]
> @rzetelskik: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:
> 
> Test name	Commit	Details	Required	Rerun command
> ci/prow/e2e-gke-parallel	[fa4b0ba](https://github.com/scylladb/scylla-operator/commit/fa4b0baab972bf2993edfa30ab5042dfbb4358f0)	[link](https://prow.scylla-operator.scylladb.com/view/gs/scylla-operator-prow/pr-logs/pull/scylladb_scylla-operator/1911/pull-scylla-operator-master-e2e-gke-parallel/1787846112095244289)	true	`/test e2e-gke-parallel`
> [Full PR test history](https://prow.scylla-operator.scylladb.com/pr-history?org=scylladb&repo=scylla-operator&pr=1911). [Your PR dashboard](https://prow.scylla-operator.scylladb.com/pr?query=is:pr+state:open+author:rzetelskik).

known manager flake
/retest

**Pull request merged**